File: loadercomplex-forge-18/src/main/java/com/enderzombi102/loadercomplex/forge18/impl/ForgeFactoryWorld.java
package com.enderzombi102.loadercomplex.forge18.impl;

import com.enderzombi102.loadercomplex.Utils;
import com.enderzombi102.loadercomplex.api.block.Blockstate;
import com.enderzombi102.loadercomplex.api.entity.Entity;
import com.enderzombi102.loadercomplex.api.entity.ItemEntity;
import com.enderzombi102.loadercomplex.api.item.ItemStack;
import com.enderzombi102.loadercomplex.api.utils.FactoryWorld;
import com.enderzombi102.loadercomplex.api.utils.ResourceIdentifier;
import com.enderzombi102.loadercomplex.api.utils.Server;
import com.enderzombi102.loadercomplex.api.world.World;
import com.enderzombi102.loadercomplex.forge18.impl.block.ForgeBlockstate;
import com.enderzombi102.loadercomplex.forge18.impl.entity.ForgeEntity;
import com.enderzombi102.loadercomplex.forge18.impl.entity.ForgeItemEntity;
import com.enderzombi102.loadercomplex.forge18.impl.item.ForgeItemStack;
import net.minecraft.block.Block;
import net.minecraft.entity.EntityType;
import net.minecraft.item.Item;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import net.minecraftforge.registries.RegistryManager;

public class ForgeFactoryWorld implements FactoryWorld {
	@Override
	public ItemStack createStack(ResourceIdentifier type) {
		Item item = RegistryManager.ACTIVE.getRegistry( Registry.ITEM_KEY )
				.getValue( new Identifier( type.getNamespace(), type.getPath() ) );
		if ( item == null )
			throw new IllegalArgumentException( Utils.format( "Item \"{}\" does not exist!", type ) );
		return new ForgeItemStack( new net.minecraft.item.ItemStack( item ) );
	}

	@Override
	public Entity createEntity(World world, ResourceIdentifier type) {
		var entityType = RegistryManager.ACTIVE.getRegistry( Registry.ENTITY_TYPE_KEY )
				.getValue( new Identifier( type.getNamespace(), type.getPath() ) );
		if ( entityType == null )
			throw new IllegalArgumentException( Utils.format( "Entity \"{}\" does not exist!", type ) );
		return new ForgeEntity( entityType.create( (net.minecraft.world.World) world.getObject() ) );
	}

	@Override
	public ItemEntity createItemEntity(World world, ItemStack stack) {
		var entity = EntityType.ITEM.create( (net.minecraft.world.World) world.getObject() );
		if ( entity == null )
			throw new IllegalArgumentException( "ItemEntity failed to spawn!" );
		entity.setStack( (net.minecraft.item.ItemStack) stack.getStack() );
		return new ForgeItemEntity( entity );
	}

	@Override
	public Blockstate createBlockstate(ResourceIdentifier type) {
		Block block = RegistryManager.ACTIVE.getRegistry( Registry.BLOCK_KEY )
				.getValue( new Identifier( type.getNamespace(), type.getPath() ) );
		if ( block == null )
			throw new IllegalArgumentException( Utils.format( "Block \"{}\" does not exist!", type ) );
		return new ForgeBlockstate( block.getDefaultState() );
	}

	@Override
	public Blockstate airBlockstate() {
		return createBlockstate( new ResourceIdentifier( "minecraft", "air" ) );
	}

	@Override
	public World adaptWorld(Server server, int id) {
		return null;
	}
}

File: loadercomplex-forge/src/ver182/kotlin/com/enderzombi102/loadercomplex/forge182/impl/platform/ForgeFactoryWorld.kt
package com.enderzombi102.loadercomplex.forge182.impl.platform

import com.enderzombi102.loadercomplex.api.math.Vec3d
import com.enderzombi102.loadercomplex.api.minecraft.block.BlockState
import com.enderzombi102.loadercomplex.api.minecraft.entity.Entity
import com.enderzombi102.loadercomplex.api.minecraft.entity.ItemEntity
import com.enderzombi102.loadercomplex.api.minecraft.item.ItemStack
import com.enderzombi102.loadercomplex.api.minecraft.server.Server
import com.enderzombi102.loadercomplex.api.minecraft.util.ResourceIdentifier
import com.enderzombi102.loadercomplex.api.minecraft.world.World
import com.enderzombi102.loadercomplex.api.platform.FactoryWorld
import com.enderzombi102.loadercomplex.forge182.impl.utils.toMC
import com.enderzombi102.loadercomplex.forge182.impl.block.ForgeBlockState
import com.enderzombi102.loadercomplex.forge182.impl.entity.ForgeEntity
import com.enderzombi102.loadercomplex.forge182.impl.entity.ForgeItemEntity
import com.enderzombi102.loadercomplex.forge182.impl.item.ForgeItemStack
import net.minecraft.entity.EntityType
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import net.minecraftforge.registries.RegistryManager
import net.minecraft.world.World as McWorld

class ForgeFactoryWorld : FactoryWorld {
	override fun createStack(type: ResourceIdentifier): ItemStack {
		val item = RegistryManager.ACTIVE.getRegistry(Registry.ITEM_KEY)
			.getValue(Identifier(type.namespace, type.path))
		requireNotNull(item) { String.format("Item \"%s\" does not exist!", type) }
		return ForgeItemStack(net.minecraft.item.ItemStack(item))
	}

	override fun createEntity(world: World, type: ResourceIdentifier, pos: Vec3d ): Entity? {
		val entityType = RegistryManager.ACTIVE.getRegistry(Registry.ENTITY_TYPE_KEY).getValue(type.toMC())
		requireNotNull(entityType) { String.format("Entity \"%s\" does not exist!", type) }
		return entityType.create(world.getObject() as McWorld)?.let { ForgeEntity(it) }
	}

	override fun createItemEntity(world: World, stack: ItemStack, pos: Vec3d ): ItemEntity {
		val entity = EntityType.ITEM.create(world.getObject() as McWorld)
		requireNotNull(entity) { "ItemEntity failed to spawn!" }
		entity.stack = stack.stack as net.minecraft.item.ItemStack
		return ForgeItemEntity(entity)
	}

	override fun createBlockstate(type: ResourceIdentifier): BlockState {
		val block = RegistryManager.ACTIVE.getRegistry(Registry.BLOCK_KEY)
			.getValue(type.toMC())
		requireNotNull(block) { String.format("Block \"%s\" does not exist!", type) }
		return ForgeBlockState(block.defaultState)
	}

	override fun airBlockstate(): BlockState {
		return createBlockstate(ResourceIdentifier("minecraft", "air"))
	}

	override fun adaptWorld(server: Server, id: Int): World? {
		throw NotImplementedError()
	}
}
